cpp-terminal/base.cpp

```diff
@@ -1,95 +1,427 @@
+    }
+    // check the color matches a 4bit color
+    if (rgb_compare(color, Bit4_reference::BLACK) < 1) {
+        return 0;
+    } else if (rgb_compare(color, Bit4_reference::RED) < 1) {
+        return 1;
+    } else if (rgb_compare(color, Bit4_reference::GREEN) < 1) {
+        return 2;
+    } else if (rgb_compare(color, Bit4_reference::YELLOW) < 1) {
+        return 3;
+    } else if (rgb_compare(color, Bit4_reference::BLUE) < 1) {
+        return 4;
+    } else if (rgb_compare(color, Bit4_reference::MAGENTA) < 1) {
+        return 5;
+    } else if (rgb_compare(color, Bit4_reference::CYAN) < 1) {
+        return 6;
+    } else if (rgb_compare(color, Bit4_reference::WHITE) < 1) {
+        return 7;
+    } else if (rgb_compare(color, Bit4_reference::GRAY) < 1) {
+        return 8;
+    } else if (rgb_compare(color, Bit4_reference::RED_BRIGHT) < 1) {
+        return 9;
+    } else if (rgb_compare(color, Bit4_reference::GREEN_BRIGHT) < 1) {
+        return 10;
+    } else if (rgb_compare(color, Bit4_reference::YELLOW_BRIGHT) < 1) {
+        return 11;
+    } else if (rgb_compare(color, Bit4_reference::BLUE_BRIGHT) < 1) {
+        return 12;
+    } else if (rgb_compare(color, Bit4_reference::MAGENTA_BRIGHT) < 1) {
+        return 13;
+    } else if (rgb_compare(color, Bit4_reference::CYAN_BRIGHT) < 1) {
+        return 14;
+    } else if (rgb_compare(color, Bit4_reference::WHITE_BRIGHT) < 1) {
+        return 15;
+    }
+
+    // check gray scale in 24 steps
+    if (color.r == color.g && color.r == color.b && color.g == color.b) {
+        return 232 + color.r / 32 + color.g / 32 + color.b / 32;
+    }
+
+    // normal color space
+    return 16 + 36 * (color.r / 51) + 6 * (color.g / 51) + (color.b / 51);
 }
 
-std::string Term::cursor_off() {
-    return "\x1b[?25l";
+bool Term::bit24_support() {
+    std::string colorterm = std::getenv("COLORTERM");
+    if (colorterm == "truecolor" || colorterm == "24bit") {
+        return true;
+    }
+    return false;
 }
 
-std::string Term::cursor_on() {
-    return "\x1b[?25h";
+std::string Term::rgb_to_bit24_auto_fg(RGB color) {
+    if (bit24_support()) {
+        return color_fg(color);
+    } else {
+        return color_fg(rgb_to_bit8(color));
+    }
 }
 
-std::string Term::clear_screen() {
-    return "\033[2J";
+std::string Term::rgb_to_bit24_auto_bg(RGB color) {
+    if (bit24_support()) {
+        return color_bg(color);
+    } else {
+        return color_bg(rgb_to_bit8(color));
+    }
 }
 
-std::string Term::clear_screen_buffer() {
-    return "\033[3J";
+/* FOREGROUND COLORS */
+
+std::string Term::color_fg(Term::Color4 color) {
+    return "\033[" + std::to_string((uint8_t)color + 30) + 'm';
 }
 
-std::string Term::move_cursor(size_t row, size_t col) {
-    return "\x1b[" + std::to_string(row) + ';' + std::to_string(col) + 'H';
+std::string Term::color_fg(uint8_t color) {
+    return "\033[38;5;" + std::to_string(color) + 'm';
```

Comment:
What happens if color is out of range?

Reply:
uint8_t can hold values from 0 to 255, which is exactly the range of the color codes. I hope that people get a compiler warning when they are doing a conversion from a bigger datatype. Any thoughts on that?